model/inline/src/main/scala/aqua/model/inline/ArrowInliner.scala

```diff
@@ -90,60 +90,55 @@ object ArrowInliner extends Logging {
-            (tree, results) = tr
+    for {
+      // Register captured values as available exports
+      _ <- Exports[S].resolved(capturedValues)
+      _ <- Mangler[S].forbid(capturedValues.keySet)
+
+      // Now, substitute the arrows that were received as function arguments
+      // Use the new op tree (args are replaced with values, names are unique & safe)
+      callableFuncBodyNoTopology <- TagInliner.handleTree(tree, funcName)
+      callableFuncBody =
+        capturedTopology
+          .fold[OpModel](SeqModel)(ApplyTopologyModel.apply)
+          .wrap(callableFuncBodyNoTopology)
+
+      opsAndRets <- pushStreamResults(
+        outsideDeclaredStreams,
+        call.exportTo,
+        results,
+        callableFuncBody
+      )
+      (ops, rets) = opsAndRets
 
-            // Register captured values as available exports
-            _ <- Exports[S].resolved(fn.capturedValues)
-            _ <- Mangler[S].forbid(fn.capturedValues.keySet)
-
-            // Now, substitute the arrows that were received as function arguments
-            // Use the new op tree (args are replaced with values, names are unique & safe)
-            callableFuncBodyNoTopology <- TagInliner.handleTree(tree, fn.funcName)
-            callableFuncBody =
-              fn.capturedTopology
-                .fold[OpModel](SeqModel)(ApplyTopologyModel.apply)
-                .wrap(callableFuncBodyNoTopology)
-
-            opsAndRets <- pushStreamResults(
-              outsideDeclaredStreams,
-              call.exportTo,
-              results,
-              callableFuncBody
-            )
-            (ops, rets) = opsAndRets
-
-            exports <- Exports[S].exports
-            arrows <- Arrows[S].arrows
-            // gather all arrows and variables from abilities
-            returnedFromAbilities = rets.collect { case VarModel(name, st @ AbilityType(_, _), _) =>
-              getVarsAndArrowsFromAbilities(name, None, st, exports, arrows)
-            }.foldMapA(_.bimap(_.toList, _.toList)).bimap(_.toMap, _.toMap)
-
-            // find and get resolved arrows if we return them from the function
-            returnedArrows = rets.collect { case VarModel(name, ArrowType(_, _), _) =>
-              name
-            }.toSet
-            arrowsToSave <- Arrows[S].pickArrows(returnedArrows)
-          } yield {
-            val (valsFromAbilities, arrowsFromAbilities) = returnedFromAbilities
-            InlineResult(
-              SeqModel.wrap(ops.reverse: _*),
-              rets.reverse,
-              valsFromAbilities,
-              arrowsFromAbilities ++ arrowsToSave
-            )
-          }
-        )
+      exports <- Exports[S].exports
+      arrows <- Arrows[S].arrows
+      // gather all arrows and variables from abilities
+      returnedFromAbilities = rets.collect { case VarModel(name, st @ AbilityType(_, _), _) =>
+        getVarsAndArrowsFromAbilities(name, None, st, exports, arrows)
```

Comment:
Why not just use `Exports[S]` and `Arrows[S]` inside `getVarsAndArrowsFromAbilities`?

Reply:
Oh, I see, this function is used somewhere else where it is harder to do so